Add rails2_ruby2 as a dependency

zdennis opened the pull request:
# What does this PR do?

This PR adds the [rails2_ruby2](https://github.com/Genius/rails2_ruby2) patches to this fork of Rails so we can ensure those patches are run against the Rails test suite. This also makes 2.3.8 the default Ruby version for this fork of rails.

## Notable things

* Updates TravisCI to build against Ruby 2.3.8 and PostgreSQL 10

* Adds pry-byebug to the Gemfile for development purposes so it's easier to debug issues when pulling in upstream changes.

zdennis committed:
Updating to Ruby 2.3.8 as the default.

zdennis committed:
Backport an activesupport duration test fix.

The prior code would fail at 3pm EST and would not pass against until midnight the next day.

Reference: https://github.com/rails/rails/commit/d81c606fbaaab1d6ad641c4c551d7b9825bae885

zdennis committed:
Adding helper scripts for running tests

To run all for Ruby 2.3 then run:

    bin/test-2.3/all

To run individual components run the corresponding bin/test-2.3/NAME_HERE script.

NOTE: the activemodel tests currently fail due to a circular dependency issue with Mocha. This was not caught earlier because the “railslts:test” rake task does not run activemodel tests.

zdennis committed:
Temporarily take out rails2_ruby2 requires.

Focus on getting travis CI matrix for Ruby 1.8.7 and 2.3.8 working.

zdennis committed:
Update TravisCI config to build run a REE 1.8.7 and Ruby 2.3.8 test matrix
with the proper Gemfiles.

This specifically excludes REE+Gemfile2.3 and 2.3.8+Gemfile.1.8 since those invalid test runs.

zdennis committed:
Install libfcgi for the fcgi gem on Ruby 2.3.8
so the fcgi gem can compile and install

zdennis committed:
TravisCI: Update Gemfile.common to use hash-rocket syntax

so it can be parsed when running REE 1.8.7

zdennis committed:
Removing Ruby 1.8 Gemfile and from TravisCI build.

The rails2_ruby2 patches require Ruby 2+ and since we’re specifically in supporting Ruby 2+ it seems like wasted effort to maintain a backport to 1.8.7.

zdennis committed:
Update mocha version to 0.9.9

This resolves a circular dependency that was present in earlier versions

although the official fix doesn’t seem to show up until 0.10.5. However 0.9.10 and higher all break mock expectations and cause additional failures.

Fix introduced in 0.9.9: https://github.com/freerange/mocha/commit/9a2ca1e5cc27d9e6a1d016592f62931e12f3913c#diff-fb508868d3e215521f9408774eaad79aL5

Official fix released in 0.10.5: https://github.com/freerange/mocha/commit/a8e27d8c1d7538e0447afda0b6ca48c5412bd0cd

zdennis committed:
Adding back in rails2_ruby2 patches

zdennis committed:
Try out travis build

Conflicts:
	.ruby-version
	Gemfile
	Gemfile.lock

zdennis committed:
Updating rack dependency to allow less stringent version requirments.

We want to allow rack 1.6.x to be used downstream.

zdennis committed:
TravisCI: Install libfcgi and libfcgi-dev so the fcgi gem

can install

zdennis committed:
Update actionpack cookie test to reflect rack updates

There are two relevant updates in rack:

First, max-age is now included in the Set-Cookie header which was released in rack 1.5.0. Reference:  https://github.com/rack/rack/commit/6779daa5646b8de27efdf0894453fb36bd36630d

Second, rack switched the format of the date/time used for expires in the Set-Cookie header, also released in rack 1.5.0. Reference: https://github.com/rack/rack/commit/5e0a9413a80e9b0aa8a74bfc806fb20f2ea0bf3c

nightpool commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -726,23 +728,41 @@ def default_sequence_name(table_name, pk = nil) #:nodoc:
+      # Based on solution for Rails 4.2 https://github.com/rails/rails/issues/28780#issuecomment-354868174
+      #
+      # Updates sequence logic to support PostgreSQL 10.
+      #
```
see also https://github.com/Genius/Rap-Genius/blob/master/config/initializers/active_record_extensions.rb#L583-L609

zdennis replied:
This has been pulled out into rails2_ruby2.

natekandler committed:
TravisCI: specify dist and postgresql packages in travis.yml

zdennis commented:
I've rebased in the latest on `2-3-genius-lts` now that PR #5 has been merged and force pushed this branch.

zdennis committed:
Update rails2_ruby2 dependency to point to the chores/postgresql_patches branch

and to update how rails2_ruby2 patches are loaded on a per-component basis.

The change in the require(s) in this commit are to allow each component to be responsible for loading their specific patches. This helps. avoid loading issues where active_support loads the rails2_ruby2 before other components are loaded (e.g. ActiveRecord) which results in the ActiveRecord patches never being loaded.

zdennis commented:
We've updated this to point to the branch for this PR: https://github.com/Genius/rails2_ruby2/pull/4 

Once that is merged in we'll update this branch to point to rails2_ruby2#master.

zdennis commented on activemodel/lib/active_model.rb:
```diff
@@ -2,4 +2,4 @@
 # require 'active_model/validations'
-require 'active_model/base'
\ No newline at end of file
+require 'active_model/base'
```
Whoops! This newline can technically go away or we can shim in the require of rails2_ruby2/active_model extensions.

zdennis commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -124,6 +124,8 @@ def self.extract_value_from_default(default)
             when /\A\(?(-?\d+(\.\d*)?\)?)\z/
               $1
+            when /\A'(-?\d+(\.\d*)?\)?)'::integer\z/i
+              $1
```
This particular patch is needed for Postgres 10.5+. Without this an ActiveRecord test fails so it felt like the right place to include this patch was in activerecord itself rather than in a patch in rails2_ruby2.

nightpool replied:
let's move this to a `rails2_ruby2` patch

zdennis commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -54,11 +54,11 @@ def test_since_and_ago_with_fractional_days
-    assert_in_delta((7 * 36).hours.ago, 1.5.weeks.ago, 1)
-    assert_in_delta((7 * 24 * 1.7).hours.ago, 1.7.weeks.ago, 1)
+    assert_equal((7 * 36).hours.ago(t), 1.5.weeks.ago(t))
+    assert_in_delta((7 * 24 * 1.7).hours.ago(t), 1.7.weeks.ago(t), 1)
```
This was a backport from a upstream fix in activesupport. See https://github.com/Genius/rails2/commit/8f71612b48a341d2579b4196fd30a43de4e267f4#diff-5711b20159cce378efc8716683a03b38 for more info.

zdennis committed:
Update the rails2_ruby2 dependnecy via 'bundle update --gemfile Gemfile.2.3 rails2_ruby2'

natekandler committed:
pointing at rails2_ruby2 1.0

natekandler committed:
update format of rails2_ruby2 dependency in gemspec

natekandler commented on bin/test-2.3/actionmailer:
```diff
@@ -0,0 +1,7 @@
+  bundle install --gemfile Gemfile.2.3 &&
+  pushd actionmailer &&
+  bundle exec --gemfile ../Gemfile.2.3 rake test ; popd
+  
```
It looks like the github linter isn't recognizing the new line at the end of these bin files for some reason. Maybe it's context specific since the ruby files are working correctly?

nightpool replied:
that's because there are two space on that line

natekandler replied:
Doh! I see them, I'll update to remove

natekandler committed:
add new line at end of file where it was missing

nightpool reviewed (approved):
looks good % moving the postgres 10 support fix into rails2_ruby2!

zdennis committed:
Updating rails2_ruby2 dependency to 1.0.2

zdennis commented:
All is green.
